parquet-variant/src/main/java/org/apache/parquet/variant/VariantValueWriter.java

```diff
@@ -0,0 +1,326 @@
+        case BINARY:
+          return primitiveTypeName == PrimitiveType.PrimitiveTypeName.BINARY && logicalType == null;
+        default:
+          return false;
+      }
+    } else if (typedValueField.getLogicalTypeAnnotation()
+        instanceof LogicalTypeAnnotation.ListLogicalTypeAnnotation) {
+      return variantType == Variant.Type.ARRAY;
+    } else {
+      return variantType == Variant.Type.OBJECT;
+    }
+  }
+
+  private void writeScalarValue(RecordConsumer recordConsumer, Variant variant, PrimitiveType type) {
+    switch (variant.getType()) {
+      case BOOLEAN:
+        recordConsumer.addBoolean(variant.getBoolean());
+        break;
+      case BYTE:
+        recordConsumer.addInteger(variant.getByte());
+        break;
+      case SHORT:
+        recordConsumer.addInteger(variant.getShort());
+        break;
+      case INT:
+        recordConsumer.addInteger(variant.getInt());
+        break;
+      case LONG:
+        recordConsumer.addLong(variant.getLong());
+        break;
+      case FLOAT:
+        recordConsumer.addFloat(variant.getFloat());
+        break;
+      case DOUBLE:
+        recordConsumer.addDouble(variant.getDouble());
+        break;
+      case DECIMAL4:
+        recordConsumer.addInteger(variant.getDecimal().unscaledValue().intValue());
+        break;
+      case DECIMAL8:
+        recordConsumer.addLong(variant.getDecimal().unscaledValue().longValue());
+        break;
+      case DECIMAL16:
+        recordConsumer.addBinary(Binary.fromConstantByteArray(
+            variant.getDecimal().unscaledValue().toByteArray()));
+        break;
+      case DATE:
+        recordConsumer.addInteger(variant.getInt());
+        break;
+      case TIME:
+        recordConsumer.addLong(variant.getLong());
+        break;
+      case TIMESTAMP_TZ:
+        recordConsumer.addLong(variant.getLong());
+        break;
+      case TIMESTAMP_NTZ:
+        recordConsumer.addLong(variant.getLong());
+        break;
+      case TIMESTAMP_NANOS_TZ:
+        recordConsumer.addLong(variant.getLong());
+        break;
+      case TIMESTAMP_NANOS_NTZ:
+        recordConsumer.addLong(variant.getLong());
+        break;
+      case STRING:
+        recordConsumer.addBinary(Binary.fromString(variant.getString()));
+        break;
+      case BINARY:
+        recordConsumer.addBinary(Binary.fromReusedByteBuffer(variant.getBinary()));
+        break;
+      default:
+        throw new IllegalArgumentException("Unsupported scalar type: " + variant.getType());
+    }
+  }
+
+  private void writeArrayValue(RecordConsumer recordConsumer, Variant variant, GroupType arrayType) {
+    if (variant.getType() != Variant.Type.ARRAY) {
+      throw new IllegalArgumentException("Expected array type but got: " + variant.getType());
+    }
+
+    // Get the element type from the array schema
+    GroupType listType = arrayType.getType(0).asGroupType();
+    Type elementType = listType.getType(0);
+
+    recordConsumer.startGroup();
```

Comment:
Comments would help here. The Avro list writer has `// group wrapper (original type LIST)`, which would probably be `// list field, annotated as LIST` now.